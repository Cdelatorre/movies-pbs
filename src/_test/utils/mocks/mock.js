export const moviesData = {
  viewedMovies: [
    {
      id: "1",
      title: "Movie 1",
      genres: ["Horror", "Comedy"],
      viewed: true,
      img: "https://picsum.photos/id/118/300/400",
      createdAt: "2022-10-16T20:25:58.420Z",
    },
  ],
  unViewedMovies: [
    {
      id: "2",
      title: "Movie 2",
      genres: ["Horror", "Comedy"],
      viewed: false,
      img: "https://picsum.photos/id/159/300/400",
      createdAt: "2022-10-16T20:51:29.081Z",
    },
    {
      id: "3",
      title: "Movie 3",
      genres: ["Romance"],
      viewed: false,
      img: "https://picsum.photos/id/122/300/400",
      createdAt: "2022-10-16T20:51:45.249Z",
    },
    {
      id: "4",
      title: "Movie 4",
      genres: ["Test"],
      viewed: false,
      img: "https://picsum.photos/id/168/300/400",
      createdAt: "2022-10-16T20:52:14.784Z",
    },
    {
      id: "5",
      title: "Movie 5",
      genres: ["Random"],
      viewed: false,
      img: "https://picsum.photos/id/149/300/400",
      createdAt: "2022-10-16T20:52:36.529Z",
    },
  ],
};

export const singleMovieData = {
  id: "1",
  title: "Movie 1",
  genres: ["Horror", "Comedy", "Romance"],
  viewed: false,
  img: "https://picsum.photos/id/149/300/400",
  createdAt: "2022-10-16T20:52:36.529Z",
};

export const activeFiltersData = {
  activeFilters: [],
  search: "",
};

export const activeLoadingData = {
  value: false,
};

export const emptyStoreData = {
  movies: { viewedMovies: [], unViewedMovies: [] },
  filters: { activeFilters: [], search: "" },
  loading: { value: false },
};

export const sampleFiltersData = ["Horror", "Comedy", "Romance"];

export const customStoreData = (movies, activeFilters, loading) => {
  return {
    movies: movies ? moviesData : emptyStoreData.movies,
    filters: activeFilters ? activeFiltersData : emptyStoreData.filters,
    loading: loading ? activeLoadingData : emptyStoreData.loading,
  };
};
