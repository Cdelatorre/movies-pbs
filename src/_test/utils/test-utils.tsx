import React, { PropsWithChildren } from "react";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import type { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { AppStore, RootState, setupStore } from "../../store";
import { BrowserRouter, Router } from "react-router-dom";
import { emptyStoreData } from "./mocks/mock";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = emptyStoreData,

    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
  route?: string
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (route) window.history.pushState({}, "Test routing", route);

    return (
      <BrowserRouter>
        <Provider store={store}>{children}</Provider>;
      </BrowserRouter>
    );
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouter(ui: React.ReactElement, history?: any) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    if (history) {
      return (
        <Router location="/" navigator={history}>
          {children}
        </Router>
      );
    } else {
      return <BrowserRouter>{children}</BrowserRouter>;
    }
  }
  return { ...render(ui, { wrapper: Wrapper }) };
}
