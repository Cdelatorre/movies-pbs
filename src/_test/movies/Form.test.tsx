import { fireEvent, screen } from "@testing-library/react";
import MovieForm from "../../components/movies/form";
import { customStoreData } from "../utils/mocks/mock";
import { renderWithProviders } from "../utils/test-utils";

describe("Renders Movie form component correctly", () => {
  test("Renders default Movie form and each tag correctly", () => {
    renderWithProviders(<MovieForm />);
    const titleInput = screen.getByTestId("input-text-title");
    const genreInput = screen.getByTestId("input-text-currentGenre");

    expect(screen.getByTestId("form-component")).toBeInTheDocument();
    expect(titleInput).toBeInTheDocument();
    expect(genreInput).toBeInTheDocument();
  });

  test("Fills inputs correctly on user input interaction", () => {
    renderWithProviders(<MovieForm />);
    const titleInput = screen.getByTestId("input-text-title");
    const genreInput = screen.getByTestId("input-text-currentGenre");

    fireEvent.change(titleInput, {
      target: { value: "Movie mock title" },
    });
    expect(screen.getByDisplayValue("Movie mock title")).toBeInTheDocument();
    fireEvent.change(genreInput, {
      target: { value: "Genre" },
    });
    expect(screen.getByDisplayValue("Genre")).toBeInTheDocument();
  });

  test("If user press enter key and genre input is focused & filled, is added as GenreTab component below title input", () => {
    renderWithProviders(<MovieForm />);
    const genreInput = screen.getByTestId("input-text-currentGenre");

    fireEvent.change(genreInput, {
      target: { value: "Genre" },
    });
    expect(screen.getByDisplayValue("Genre")).toBeInTheDocument();

    genreInput.focus();

    fireEvent.keyPress(genreInput, { key: "Enter", keyCode: 13, charCode: 13 });

    expect(screen.getByTestId("genre-tab-Genre")).toBeInTheDocument();
    expect(screen.queryByDisplayValue("Genre")).toBeNull();
  });

  test("Create movie in form calls redux dispatch", () => {
    const customStore = customStoreData(true, false, false);

    const { store } = renderWithProviders(<MovieForm />, {
      preloadedState: {
        ...customStore,
      },
    });

    console.log(1, store.getState().movies.unViewedMovies);

    const titleInput = screen.getByTestId("input-text-title");
    const genreInput = screen.getByTestId("input-text-currentGenre");
    const submitButton = screen.getByText("Create");

    /*---- filling form ----*/
    fireEvent.change(titleInput, {
      target: { value: "Movie mock title" },
    });
    fireEvent.change(genreInput, {
      target: { value: "Genre" },
    });
    genreInput.focus();
    fireEvent.keyPress(genreInput, { key: "Enter", keyCode: 13, charCode: 13 });
    /*----------------------*/

    fireEvent.click(submitButton);
    expect(screen.queryByDisplayValue("Genre")).toBeNull();
    expect(screen.queryByDisplayValue("Movie mock title")).toBeNull();

    expect(
      store
        .getState()
        .movies.unViewedMovies.find(
          (movie) =>
            movie.title === "Movie mock title" && movie.genres.includes("Genre")
        )
    ).toBeTruthy();
  });
});
