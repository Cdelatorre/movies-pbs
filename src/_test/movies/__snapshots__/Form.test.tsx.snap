// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Movie form component correctly Renders default Movie form and each tag correctly 1`] = `
<div
  class="movie-form-container mt-3"
  data-testid="form-component"
>
  <h2
    class="movie-form-title mt-0"
  >
    Create your movie
  </h2>
  <form
    class="movie-form"
    data-testid="movie-form"
  >
    <div
      class="input "
    >
      <label
        class="form-label w-100 text-start text-light"
        for="title-input"
      >
        Title
      </label>
      <input
        class="form-control form-text bg-dark text-light"
        data-testid="input-text-title"
        id="title-input"
        name="title"
        placeholder="Write down a title for your movie"
        type="text"
        value=""
      />
    </div>
    <div
      class="genders d-flex mb-3 mt-3"
    />
    <div
      class="input "
    >
      <label
        class="form-label w-100 text-start text-light"
        for="title-input"
      >
        Genres
      </label>
      <input
        class="form-control form-text bg-dark text-light"
        data-testid="input-text-currentGenre"
        id="title-input"
        name="currentGenre"
        placeholder="Write down some movie genres"
        type="text"
        value=""
      />
    </div>
  </form>
  <div
    class="mt-4"
  >
    <button
      class="btn btn-success w-100 mt-3"
      data-testid="test-submit-form-btn"
      id="submit-form-btn"
    >
      Create
    </button>
  </div>
</div>
`;
